pkg/storage/etcd/etcd_watcher.go

```diff
@@ -144,10 +144,25 @@ func (w *etcdWatcher) etcdWatch(ctx context.Context, client etcd.KeysAPI, key st
 	defer util.HandleCrash()
 	defer close(w.etcdError)
 	defer close(w.etcdIncoming)
+
+	// We need to be prepared, that Stop() can be called at any time.
+	// It can potentially also be called, even before this function is called.
+	// If that is the case, we simply skip all the code here.
+	// See #18928 for more details.
+	w.stopLock.Lock()
+	if w.stopped {
+		// Watcher has already been stopped - don't event initiate it here.
+		w.stopLock.Unlock()
```

Comment:
Can you use an anonymous function or a separate function to encapsulate this logic before the infinite for loop so you can just defer Unlock? I'd hate for someone to come add another error condition with an early return and forget to unlock.